lib/Conversion/ConvertToArcs/ConvertToArcs.cpp

```diff
@@ -35,15 +35,14 @@ static LogicalResult convertInitialValue(seq::CompRegOp reg,
   if (!reg.getInitialValue())
     return values.push_back({}), success();
 
-  auto init = circt::seq::unwrapImmutableValue(reg.getInitialValue());
-  if (!init.getDefiningOp<hw::ConstantOp>())
-    return reg.emitError() << "non-constant initial value not supported";
-
-  // Clone the initial value to the top-level.
-  auto *op = init.getDefiningOp()->clone();
-  reg->getBlock()->getOperations().insert(Block::iterator(reg), op);
-  auto result = op->getResult(0);
-  values.push_back(result);
+  // unrealized_conversion_cast to normal type
+  OpBuilder builder(reg);
+  auto init = builder
+                  .create<mlir::UnrealizedConversionCastOp>(
+                      reg.getLoc(), reg.getType(), reg.getInitialValue())
+                  .getResult(0);
```

Comment:
I'm wondering if it might be worthwhile having an `immutable<T> -> T` cast operation, something like `seq.unwrap_immutable`. That could also enable some canonicalizations where a `seq.compreg` could lower to its initial value if it provably can never change.

Unrealized conversion cast also works though :smiley:

Reply:
Or we might reconsider supporting `arc.initial` before lower state and do the conversion to it in ConvertToArcs?

Reply:
`seq.unwrap_immutable` sounds reasonable and will follow-up. `arc.initial` might work but currently `arc.initial` doesn't produce any value and we need to allocate states for them. So I guess there would be a big overlap with LowerState. Another approach might be to change `arc.state`'s initial value operand to take `seq.immutable` type instead of a normal integer.